sklearn/cluster/_k_means.pyx

```diff
@@ -8,167 +8,207 @@
 
-    Return the inertia (sum of squared distances to the centers).
+cpdef floating _inertia_sparse(X,
+                               floating[::1] sample_weight,
+                               floating[:, ::1] centers,
+                               int[::1] labels):
+    """Compute inertia for sparse input data
+    
+    Sum of squared distance between each sample and it's assigned center.
     """
     cdef:
-        unsigned int n_clusters = centers.shape[0]
-        unsigned int n_features = centers.shape[1]
-        unsigned int n_samples = X.shape[0]
-        unsigned int x_stride
-        unsigned int center_stride
-        unsigned int sample_idx, center_idx, feature_idx
-        unsigned int store_distances = 0
-        unsigned int k
-        np.ndarray[floating, ndim=1] center_squared_norms
-        # the following variables are always double cause make them floating
-        # does not save any memory, but makes the code much bigger
-        DOUBLE inertia = 0.0
-        DOUBLE min_dist
-        DOUBLE dist
-
-    if floating is float:
-        center_squared_norms = np.zeros(n_clusters, dtype=np.float32)
-        x_stride = X.strides[1] / sizeof(float)
-        center_stride = centers.strides[1] / sizeof(float)
-        dot = sdot
-    else:
-        center_squared_norms = np.zeros(n_clusters, dtype=np.float64)
-        x_stride = X.strides[1] / sizeof(DOUBLE)
-        center_stride = centers.strides[1] / sizeof(DOUBLE)
-        dot = ddot
-
-    if n_samples == distances.shape[0]:
-        store_distances = 1
+        floating[::1] X_data = X.data
+        int[::1] X_indices = X.indices
+        int[::1] X_indptr = X.indptr
 
-    for center_idx in range(n_clusters):
-        center_squared_norms[center_idx] = dot(
-            n_features, &centers[center_idx, 0], center_stride,
-            &centers[center_idx, 0], center_stride)
-
-    for sample_idx in range(n_samples):
-        min_dist = -1
-        for center_idx in range(n_clusters):
-            dist = 0.0
-            # hardcoded: minimize euclidean distance to cluster center:
-            # ||a - b||^2 = ||a||^2 + ||b||^2 -2 <a, b>
-            dist += dot(n_features, &X[sample_idx, 0], x_stride,
-                        &centers[center_idx, 0], center_stride)
-            dist *= -2
-            dist += center_squared_norms[center_idx]
-            dist += x_squared_norms[sample_idx]
-            dist *= sample_weight[sample_idx]
-            if min_dist == -1 or dist < min_dist:
-                min_dist = dist
-                labels[sample_idx] = center_idx
-
-        if store_distances:
-            distances[sample_idx] = min_dist
-        inertia += min_dist
+        int n_samples = X_indptr.shape[0] - 1
+        int n_features = centers.shape[1]
+        int i, j, k
+        int row_ptr, nz_len, nz_ptr
+        floating tmp, sample_inertia
+
+        floating inertia = 0.0
+
+    for i in range(n_samples):
```

Comment:
what about `enumerate()` ? then you may drop also `j = labels[i]`

Reply:
If I used an enumerate it would be a python loop which would be much slower.